### PR TITLE
Introduce Option type

### LINKED ISSUE #4: Option
https://github.com/scala/scala/blob/2.12.x/src/library/scala/Option.scala

### PR DESCRIPTION
**Pull request by pawaclawczyk**
Resolve #4

### CODE REVIEW

File: src/Scalp/functions.php [hunk 1]
```diff
@@ -22,6 +22,23 @@ function AnyToString($any): string
     }
 }
 
+namespace Scalp {
+    function None()
```
Comment (lzakrzewski) [inline src/Scalp/functions.php]: Return type declaration is missing

File: src/Scalp/Option.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scalp;
+
+use function Scalp\Type\restrictCallableReturnType;
+
+abstract class Option
```
Comment (lzakrzewski) [inline src/Scalp/Option.php]: Readme was not updated.

File: tests/Scalp/SomeTest.php [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    /** @test */
+    public function map_applies_function_and_returns_result_wrapped_in_some(): void
+    {
+        $this->assertEquals(Some(1764), Some(42)->map(function (int $x): int { return $x ** 2; }));
+    }
+
+    /** @test */
+    public function fold_applies_function_ant_returns_result(): void
```
Comment (pawelbaranski) [inline tests/Scalp/SomeTest.php]: typo `ant`

File: tests/Scalp/SomeTest.php [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+        $f = function (int $x): int {
+            return $x ** 2;
+        };
+
+        $this->assertEquals(1764, Some(42)->fold($ifEmpty, $f));
+    }
+
+    /** @test */
+    public function flat_map_applies_function_ant_returns_result(): void
```
Comment (pawelbaranski) [inline tests/Scalp/SomeTest.php]: typo `ant`

**Pull request merged**